### ISSUE TITLE
samtools merge -c and -p options not working as expected?

### ISSUE THREAD

**Issue by mcshane** [samtools/samtools#411]
samtools merge on a set of bam files gives these readgroups:
```
>samtools merge -u - ~sm15/merge_bug/*.bam | samtools view -H | grep ^@RG
@RG	ID:2822_6	LB:LIB01	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
@RG	ID:2822_6-484275CC	LB:LIB01	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
@RG	ID:2822_6-27339861	LB:LIB01	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
@RG	ID:2822_7	LB:LIB01	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
@RG	ID:2822_7-17F89869	LB:LIB01	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
@RG	ID:2823_4	LB:LIB02	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
@RG	ID:2823_4-7A173AF0	LB:LIB02	SM:SAMPLE01	CN:SC	PL:ILLUMINA	DS:STUDY01
```
Fine. But wanted to use the `-c` option to collapse readgroups with the same ID rather than appending a random string. Using `-c` though removes all `@RG` lines:
```
>samtools merge -cu - ~sm15/merge_bug/*.bam | samtools view -H | grep ^@RG
```

Similar problem with the `-p` option and the `@PG` tags.

**Comment by kirkmcclure** [comment]
The bug was introduced with bam_sort.c commit 55c852f (May 21, 2015) -
````
./samtools merge -cu - test/dat/test_input_1_a.bam test/dat/test_input_1_c.bam
...
@SQ	SN:ref3	LN:4
@PG	ID:bull	PP:donkey
...
````
Previous version of bam_sort.c - e8f27a3 (May 12, 2015) -
````
./samtools merge -cu - test/dat/test_input_1_a.bam test/dat/test_input_1_c.bam
...
@SQ	SN:ref3	LN:4
@RG	ID:fish	PG:donkey
@RG	ID:cow	PU:13_&^&&*(:332
@RG	PU:*9u8jkjjkjd:	ID:colt
@PG	ID:bull	PP:donkey
...
````

**Comment by SamStudio8** [comment]
I'm guessing there is not a test-case that checks for RG/PG merges, then?

**Comment by kirkmcclure** [comment]
Restoring the previous initialization code gives the expected result - replace
````
        trans_tbl_init(hout, hin, translation_tbl+i, flag & MERGE_COMBINE_RG, flag & MERGE_COMBINE_PG, RG[i]);
````
with
````
        if (hout && hout->l_text>0)
            trans_tbl_init(hout, hin, translation_tbl+i, flag & MERGE_COMBINE_RG, flag & MERGE_COMBINE_PG, RG[i]);
        else {
            if (hout) bam_hdr_destroy(hout);
            // As yet, no headers to merge into...
            hout = bam_hdr_dup(hin);
            // ...so no need to translate header into itself
            trans_tbl_init(hout, hin, translation_tbl+i, true, true, NULL);
        }
````
then
````
./samtools merge -cu - test/dat/test_input_1_a.bam test/dat/test_input_1_c.bam
...
@RG	ID:fish	PG:donkey
@RG	ID:cow	PU:13_&^&&*(:332
@RG	PU:*9u8jkjjkjd:	ID:colt
...

./samtools merge -u - test/dat/test_input_1_a.bam test/dat/test_input_1_c.bam
...
@RG	ID:fish	PG:donkey
@RG	ID:cow	PU:13_&^&&*(:332
@RG	PU:*9u8jkjjkjd:	ID:colt
@RG	ID:fish-71E76AEF
@RG	ID:cow-1D2D1EE4	PU:13_&^&&*(:332
@RG	PU:*9u8jkjjkjd:	ID:colt-4C220D28
...
````
Incidentally, setup_test_1() etc. in test_trans_tbl_init.c similarly precondition the output string.

**Comment by mp15** [comment]
Fixed by 227c142c288b4757e9bd7907a0b52a9f16ee451b
